arangod/IResearch/IResearchAnalyzerFeature.cpp

```diff
@@ -1644,6 +1716,231 @@ Result IResearchAnalyzerFeature::emplace(
+  auto analyzerModificationTrx = createAnalyzerModificationTransaction(server(), vocbase.name());
+  if (analyzerModificationTrx) {
+    auto startRes = analyzerModificationTrx->start();
+    if (startRes.fail()) {
+      return startRes;
+    }
+    auto cleanupResult = cleanupAnalyzersCollection(vocbase.name(),
+                                                    analyzerModificationTrx->buildingRevision());
+    if (cleanupResult.fail()) {
+      return cleanupResult;
+    }
+  }
+  auto* engine = EngineSelectorFeature::ENGINE;
+  TRI_ASSERT(engine && !engine->inRecovery());
+  if (!analyzerModificationTrx) {
+    // no modification transaction. Just truncate
+    auto ctx = transaction::StandaloneContext::Create(vocbase);
+    SingleCollectionTransaction trx(ctx, arangodb::StaticStrings::AnalyzersCollection,
+                                    AccessMode::Type::EXCLUSIVE);
+
+    auto res = trx.begin();
+    if (res.fail()) {
+      return res;
+    }
+
+    OperationOptions options;
+    trx.truncateAsync(arangodb::StaticStrings::AnalyzersCollection, options).get();
+    res = trx.commitAsync().get();
+    if (res.ok()) {
+      invalidate(vocbase);
+    }
+    return res;
+  } else {
+    // as in single analyzer case. First mark for deletion (but all at once)
+    auto stringRev = std::to_string(analyzerModificationTrx->buildingRevision());
+    std::string aql("FOR u IN ");
+    aql.append(arangodb::StaticStrings::AnalyzersCollection).append(" UPDATE u.")
+       .append(arangodb::StaticStrings::KeyString).append(" WITH { ")
+       .append(arangodb::StaticStrings::AnalyzersDeletedRevision)
+       .append(": ").append(stringRev).append("} IN ")
+       .append(arangodb::StaticStrings::AnalyzersCollection);
+
+    auto ctx = transaction::StandaloneContext::Create(vocbase);
+    { // SingleCollectionTransaction scope
+      SingleCollectionTransaction trx(ctx, arangodb::StaticStrings::AnalyzersCollection,
+                                      AccessMode::Type::EXCLUSIVE);
+
+      auto res = trx.begin();
+      if (res.fail()) {
+        return res;
+      }
+      arangodb::aql::Query query(ctx, arangodb::aql::QueryString(aql), nullptr, nullptr);
+      aql::QueryResult queryResult = query.executeSync();
+      if (queryResult.fail()) {
+        return queryResult.result;
+      }
+      res = trx.commitAsync().get();
+      if (!res.ok()) {
+        return res;
+      }
+      res = analyzerModificationTrx->commit();
+      if (res.fail()) {
+        return res;
+      }
+    }
+    { // SingleCollectionTransaction scope
+      // now let`s do cleanup
+      SingleCollectionTransaction truncateTrx(ctx, arangodb::StaticStrings::AnalyzersCollection,
+                                              AccessMode::Type::EXCLUSIVE);
+
+      auto res = truncateTrx.begin();
+
+      if (res.ok()) {
+        OperationOptions options;
+        truncateTrx.truncateAsync(arangodb::StaticStrings::AnalyzersCollection, options).get();
```

Comment:
Yet another question: this truncates the `_analyzers` collection, but a few lines above there is an UPDATE query executes on it.
Wouldn't it be easier to only run truncate and get rid of the UPDATE query?
As far as I can see, everything that is updated will be removed by the truncate directly after anyway. Or am I missing something?

Reply:
You are missing analyzerModificationTransaction commit which is agency operation. The idea is to first mark 
analyzers as deleted starting with some revision number. And than commit that number into agency. If agency commit fails, we a save as analyzers still present and delete mark will eventually be cleared by cleanup operations. And only after successful commit we actually remove effectively dead analyzers. Even if removal fails it is ok. Leftovers will be eventually cleaned by cleanup. So why there 2 step operation.

Reply:
Got it, thanks for the explanation!